torch/csrc/autograd/engine.cpp

```diff
@@ -59,6 +59,7 @@ auto Engine::backward(const variable_list& variables,
                       tensor_list& grad_variables,
                       bool retain_variables) -> void {
   function_queue creators;
+  std::unordered_map<std::shared_ptr<Function>, GradBuffer*> creator_grad;
```

Comment:
Right now they'll be added to the `ready` queue multiple times, with different grads. After this change they're first gathered in that map, with duplicate entries having their grad accumulated, and only then are they added to the queue. You can see that I removed a `+ 0` hack in nn tests that was necessary because of the bug.